File: Common/src/main/java/at/petrak/hexcasting/api/casting/eval/sideeffects/EvalSound.java
package at.petrak.hexcasting.api.casting.eval.sideeffects;

import net.minecraft.sounds.SoundEvent;
import org.jetbrains.annotations.Nullable;

/**
 * The kind of sound that plays after a cast.
 *
 * @param sound    the actual sound file
 * @param priority the priority of this sound. the sound with the highest priority in a given cast will be
 *                 played.
 *                 shortcutMetacasting takes precedence over this.
 */
public record EvalSound(@Nullable SoundEvent sound, int priority) {
    public EvalSound greaterOf(EvalSound that) {
        return (this.priority > that.priority) ? this : that;
    }
}

File: Common/src/main/java/at/petrak/hexcasting/api/spell/casting/EvalSound.kt
package at.petrak.hexcasting.api.spell.casting

import at.petrak.hexcasting.common.lib.HexSounds
import net.minecraft.sounds.SoundEvent

/**
 * Sound that plays as a side-effect-adjacent when casting
 */
enum class EvalSound {
    /** Silence */
    NONE,

    /** The generic "bvwonh" sound for non-spell ops */
    GENERIC,

    /** The "bwoink!" for spell ops */
    SPELL_BOINK,

    /** The "gvwoh" for mishaps */
    MISHAP;

    /**
     * Which sound type has the greater priority?
     */
    fun greaterOf(that: EvalSound): EvalSound =
        EvalSound.values()[maxOf(this.ordinal, that.ordinal)]

    fun soundEvent(): SoundEvent? = when (this) {
        NONE -> null
        GENERIC -> HexSounds.ADD_PATTERN
        SPELL_BOINK -> HexSounds.ACTUALLY_CAST
        MISHAP -> HexSounds.FAIL_PATTERN
    }
}